### PR TITLE
Add reset command for nitrokey 3

### PR DESCRIPTION
**Pull request by sosthene-nitrokey**
Close https://github.com/Nitrokey/nitrokey-hotp-verification/issues/42

```
./hotp_verification reset
```

Is a noop on a device other than NK3. On an NK3, reset the secrets app and thus the HOTP code.

### CODE REVIEW

**Comment by nestire** [comment]
I get this with a nitrokey pro attached, same binary works with an nk3 
```
-rwxr-xr-x 1 user user 55K Nov 27 10:54 hotp_verification
sha256sum hotp_verification
86c94ed1bc0cfb43b3dc2b423b74a56ef80f356fc0aa83915a82ef04651c2862  hotp_verification
[user@work nitrokey-hotp-verification (nk3-reset)]$ ./hotp_verification reset
HOTP code verification application, version 1.6

Segmentation fault (core dumped)
[user@work nitrokey-hotp-verification (nk3-reset)]$ 
`````

**Comment by tlaurion** [comment]
@sosthene-nitrokey ping https://github.com/linuxboot/heads/issues/1866

**Comment by tlaurion** [comment]
> Ideally, `hotp_verification reset` would accept a parameter here, which would be the secret app PIN.
> 
> Otherwise as can be seen under https://github.com/linuxboot/heads/pull/1850/commits/07f3710f7847c07b29da249cc36c810c41ea7124, it requires an additional step which would be PIN change, but we don't have a PIN here, since we just reset :)
> 
> 
> I still think, as said under https://github.com/Nitrokey/nitrokey-hotp-verification/issues/36#issuecomment-2479284125 that changing PIN is not really important to Heads use case since we reset. What is important is setting a secret app PIN at moment of oem factory reset/re-ownership.
> 
> 
> Need:
> 
> - [ ] `hotp_verification reset DESIRED_SECRET_APP_PIN`
> 
> _Originally posted by @tlaurion in https://github.com/Nitrokey/nitrokey-hotp-verification/issues/42#issuecomment-2506814366_
>

**Comment by sosthene-nitrokey** [comment]
Yes, I'm on it, just waiting for the devices I need to reproduce and debug the segfaults, they're on their way.

**Comment by tlaurion** [comment]
> > Ideally, `hotp_verification reset` would accept a parameter here, which would be the secret app PIN.
> > 
> > Otherwise as can be seen under https://github.com/linuxboot/heads/pull/1850/commits/07f3710f7847c07b29da249cc36c810c41ea7124, it requires an additional step which would be PIN change, but we don't have a PIN here, since we just reset :)
> > 
> > 
> > I still think, as said under https://github.com/Nitrokey/nitrokey-hotp-verification/issues/36#issuecomment-2479284125 that changing PIN is not really important to Heads use case since we reset. What is important is setting a secret app PIN at moment of oem factory reset/re-ownership.
> > 
> > 
> > Need:
> > 
> > - [ ] `hotp_verification reset DESIRED_SECRET_APP_PIN`
> > 
> > _Originally posted by @tlaurion in https://github.com/Nitrokey/nitrokey-hotp-verification/issues/42#issuecomment-2506814366_
> >             

@sosthene-nitrokey ack please.

**Comment by tlaurion** [comment]
Reasoning exerpt:
> TLDR...... hotp-verification _should_
> - Either set a default secret app pin and offer pin change so re-ownership can change it just like before as part of re-ownership for <nk3 (GPG Admin PIN for <nk3 is nk3 secret app PIN for nk3. Regression as of today, best practices not followed, reinventing the wheel with weaker process was chosen as of today)
> - have `hotp_verification reset SECRET_APP_PIN` requiring a pin if none set by default
> - Have `hotp_verification reset` set a default PIN, which if we don't plan to reinvent the wheel should be equivalent to gpg Admin PIN which is 12345678.
> 
> _Originally posted by @tlaurion in https://github.com/linuxboot/heads/issues/1866#issuecomment-2508703405_

Its implementation choices. Second option chosen here, either is good, but one needed to be implemented. Note that physical presence still needed, but once instead of twice (changing PIN would also require physical presence, right? until #41 implemented in firmware version bump).

As soon as a commit provided without regression, I can easily adapt WiP code under https://github.com/linuxboot/heads/pull/1850, specifically part related to present PR change under nk3 related secret app reset needed at https://github.com/linuxboot/heads/pull/1850/files#diff-bec315d56e315e18809e58210db23903cfa07559d297a1910f2c3151f6e6e97dR141-R149

File: src/main.c [hunk 1]
```diff
@@ -38,9 +38,10 @@ void print_help(char *app_name) {
            "\t%s info\n"
            "\t%s version\n"
            "\t%s check <HOTP CODE>\n"
-           "\t%s regenerate <ADMIN PIN>\n"
+           "\t%s reset [ADMIN PIN]\n"
```
[the remark this answers is not in the document]
Comment (sosthene-nitrokey) [reply src/main.c:41]: You mean `[SECRET APP PIN]` ?

[the remark this answers is not in the document]
Comment (tlaurion) [reply src/main.c:41]: > You mean `[SECRET APP PIN]` ?

I restate that UX with many Admin PIN if no User PIN counterpart is misleading.

Yes, Secret App PIN. At least for everything exposed to user and Heads as output.

Admin /User PIN makes sense when one have power over the other which is not the case here and users continuously lost with which PIN is asked here for security dongle, where Heads can only reuse PIN to simplify things.

Secret app pin has counter of 6 for that reason as opposed to user/Admin PIN of gpg if I understood correctly.

Do you understand what I'm talking about for sake of less support requests and UX consistency and less user confusion?

Yes, secret app PIN.

[the remark this answers is not in the document]
Comment (tlaurion) [reply src/main.c:41]: > I repeat
> 
> You proposed :
> > See also issue #36 
> > Change this:
> > ```
> > 
> > HOTP code verification application, version 1.6
> > Connected device status:
> > 	Card serial: 0x7BE66C6C
> > 	Firmware: v4.13
> > 	Card counters: Admin 6, User 6
> > Operation success
> > ```
> > To
> > ```
> > HOTP code verification application, version 1.6
> > Connected device status:
> > 	Card serial: 0x7BE66C6C
> >         Firmware Nitrokey 3: v1.7.1
> > 	Firmware Secrets app: v4.13
> > 	Secret app pin counters : Admin 6, User 6
> > Operation success
> > ```
> 
> I proposed and restated:
> 
> **Even more sensical**: no secret app even named  anywhere because there is none on non nk3(regression), so no version of non existing secret app, no secret app pin, just real information :
> ```
> HOTP code verification application, version 1.7
> Connected device status:
> Card serial: 0x7BE66C6C
>          Firmware Nitrokey 2: v1.7.1
> OpenPGP smartcard PIN counters : Admin: 3, User: 3
> Operation success
> ```
> 
> For nk3:
> ```
> HOTP code verification application, version 1.7
> Connected device status:
> Card serial: 0x7BE66C6C
>          Firmware Nitrokey 3: v1.7.1
>          Firmware Secrets app: v4.13
> Secret app PIN counter : 6
> OpenPGP smartcard PIN counters : Admin: 3, User: 3
> Operation success
> ```
> 
> _Originally posted by @tlaurion in https://github.com/Nitrokey/nitrokey-hotp-verification/issues/38#issuecomment-2487450702_

**Comment by tlaurion** [comment]
I tested this commit and can say that it works in PoC at https://github.com/linuxboot/heads/pull/1850#issuecomment-2521534705

seal-hotp prompts for either Secure App PIN/GPG Admin PIN as per referred commit. 

reset works with SECURE_APP_PIN per re-ownership/oem factory reset.

Comments welcome over there. All tests can be done under qemu.

**Review by tlaurion — APPROVED** [review]
LGTM tested functionally under https://github.com/linuxboot/heads/pull/1850

**Comment by tlaurion** [comment]
@sosthene-nitrokey https://github.com/tlaurion/heads/commit/94565dc68b98a35200bcfe608cb1d256c42946dd tested with #43 and #46

Some notes though:
- "touched received" is always obtained even if no touched happened in output. It's kinda hard to help user troubleshoot what happens if the output is lying. I guess you should open an issue after replicating.
- oem-factory-reset was modified to loop 3 times when error code 3 is given by hotp-verification, even though the error code 1 is outputted on screen (inconsistence). You should probably open another issue after replicating.

Otherwise, oem-factory-reset and seal-hotp were modified. secrets App PIN is attempted automatically based on hotp-verification info output and user is warned to do re-ownership since a default PIN is used for remote attestation, defeating remote attestation purpose.

**Review by robin-nitrokey — APPROVED** [review]

**Comment by tlaurion** [comment]
> @sosthene-nitrokey https://github.com/tlaurion/heads/commit/94565dc68b98a35200bcfe608cb1d256c42946dd tested with #43 and #46
> 
> Some notes though:
> - "touched received" is always obtained even if no touched happened in output. It's kinda hard to help user troubleshoot what happens if the output is lying. I guess you should open an issue after replicating.
> - oem-factory-reset was modified to loop 3 times when error code 3 is given by `hotp-verification reset $PIN`, even though the error code 1 is outputted on screen (inconsistence). You should probably open another issue after replicating, but code takes for granted error 3 is abensce of touch/physical presence as of now, errors and push user/oem to be attentive to lack of https://github.com/Nitrokey/nitrokey-hotp-verification/issues/41 implementation
> 
> Otherwise, oem-factory-reset and seal-hotp were modified under Heads and i'm cleaning commits. Secrets App PIN is attempted automatically based on hotp-verification info output under seal-hotp and user is warned to do re-ownership since a default PIN is used for remote attestation, defeating remote attestation purpose.
> 
> TODO:
> - [ ] merge this PR
> - [ ] create additional issues/pr to fix inconsistencies of output (Heads says to contact Nitrokey support if they can't make sense of output incosistence in code if you want to review)
> - [ ] version bump hotp-verification
> - [ ] communicate with me asap so Heads upstream contains hotp-verification version bump Nitrokey is ok dealing with support requests coming their way if prior issues unfixed.
> 
> 
> Thanks @sosthene-nitrokey 
> 

@sosthene-nitrokey this PR was approved. What is preventing to hit the merge button and version bump hotp-verification version, as well as opening other issues from cited points here?

**Comment by sosthene-nitrokey** [comment]
We could just remove the `touch received` display. It's not trivial to distinguish errors that come from touch from other errors since there is no standard CCID status code for this, so devices use their own status codes, that are sometimes indistinguishable from other errors. @tlaurion

**Comment by tlaurion** [comment]
> We could just remove the `touch received` display. It's not trivial to distinguish errors that come from touch from other errors since there is no standard CCID status code for this, so devices use their own status codes, that are sometimes indistinguishable from other errors. @tlaurion

@sosthene-nitrokey : I will use error 3 returned as said. Problem I explained is that inside of hotp-verification, its showed error 1, not 3. Anyway, Will redirect support requests to nitrokey if I receive any, as in code you can review and test yourself. 

Will wait for hotp-verification version bump to revisit this. I was told that the 30h I worked on the testing/fixing Heads was not going to be paid seperately then profit haring aggreement. Work trace at https://github.com/linuxboot/heads/issues/1866 

CC: @jan23 donations expected.

**Comment by tlaurion** [comment]
@sosthene-nitrokey screenshot says more then words


- First touch received: is a lie. Heads opens whiptail window telling user 1/3 attempt to bring his attention on physical presence needed (that is error 3 rerutned, here showed per hotp-verification: 1: unknown error. Interesting, to say the least, no?
- Then nk3 says two touch received, each of one saying the truth but no clue why sometimes it goes to threee captions like this before succeeding.

I leave you with opening distinct issues or not, I will use hotp-verification version 1.7 that was just bumped and will let you do your own things until I receive a call from @jans23 for proper compensation: my job is not to fix Nitrokey stuff. It should be Nitrokey's job, unfrotunately i'm the one seeing them and collaboration was once again long, uneasy and the outcome as of today is that nk3 is still requiring physical presence so I will need to reiterate on that because ecosystem doesn't want physical presnece under Heads. So, even if i'm told to not work for free, I actually am, and I intend to put a stop on this unless i'm paid for work actually done.
